OOP proto class

JuliaLuccio opened the pull request.

JuliaLuccio committed:
teacher, student, fakeStudent

JuliaLuccio committed:
add method findFakeStudentByAverageMark

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,202 @@
+      }
+
+      getListOfNamesByAverageMark() {
+         const sortbyAvMark = () => (a, b) => a.getAverageMark() < b.getAverageMark() ? 1 : -1;
+         const copyGroup = [...this.group];
+         return copyGroup.sort(sortbyAvMark()).map(student => student.name);
```
Please use the chain way. `return [...this.group].sort((a, b) => ...).map(...)`

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,202 @@
+      }
+      
+      getStudentByName(name) {
+         const filterByName = this.group.filter(student => student.name === name);
+         return filterByName[0];
```
It's better to use `find` instead

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,202 @@
+      findFakeStudentByProto() {
+         let fakeStudent;
+         this.group.forEach(student => {
+            if(Object.getPrototypeOf(student) !== Human) {
```
It's not going to work correctly. Use `instanceof` to check whether student is `FakeStudent`

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,202 @@
+      }
+   }
+
+   const avgMark = (arr) => {
```
Do not use global data in modules/functions/classes

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,202 @@
+
+         const setMark = (mark) => {
+            const multMark = mark * multiplier;
+            return multMark >= maxMark ? preferMark : multMark;
```
Why do you need `preferMark`? Accordingly to hometask requirements it should be `maxMark`

JuliaLuccio committed:
fix

global data, chain way, find instead filter, instanceof instead getPrototypeOf, del preferMark

JuliaLuccio committed:
endline

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,199 @@
+
+      findFakeStudentByProto() {
+         let fakeStudent;
+         this.group.forEach(student => {
```
you could use `find` here as well

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,199 @@
+         [this.name, this.surname] = fullName.split(' ');
+      }
+
+      static avgMark(arr) {
```
static should be declared at the start of the class

vadkuze commented on Lesson_20/oop-principles.js:
```diff
@@ -0,0 +1,199 @@
+      }
+
+      getAverageMark() {
+         return Human.avgMark(this.marks);
```
or `Student.avgMark`. The static methods are inheritable as well

JuliaLuccio committed:
`foreach` replace to `find`

JuliaLuccio committed:
add OOP, prototype, class in README
